### PR TITLE
Migrate SLi-Rec from TensorFlow to PyTorch

### PR DESCRIPTION
**Pull request by miguelgfierro**
## Description

Migrates the **SLi-Rec** sequential recommender from TensorFlow (TF-1.x graph mode) to **PyTorch**, as part of the TF→PyTorch migration tracked in https://github.com/recommenders-team/recommenders/issues/2302.

The port is a self-contained `nn.Module` mirroring the existing PyTorch `graphrec/lightgcn.py` and the `lightgcn_deep_dive` notebook: **all hyper-parameters are explicit arguments — no `prepare_hparams`/yaml/`HParams` object**. Architecture goes on the constructor, training knobs on `fit`, evaluation knobs on `run_eval`; it reuses only the framework-agnostic metric functions in `deeprec_utils`. The **TensorFlow SLi-Rec is removed** and replaced by this port; the five other TF sequential models (asvd, caser, gru, nextitnet, sum) and the shared TF base are untouched.

```python
model = SLI_RECModel(
    user_vocab=user_vocab, item_vocab=item_vocab, cate_vocab=cate_vocab,
    item_embedding_dim=32, cate_embedding_dim=8, user_embedding_dim=16,
    hidden_size=40, attention_size=40, max_seq_length=50,
    layer_sizes=[100, 64], att_fcn_layer_sizes=[80, 40], dropout=[0.3, 0.3], seed=SEED,
)
model.fit(train_file, valid_file, epochs=10, batch_size=400, learning_rate=0.001,
          train_num_ngs=4, valid_num_ngs=4, embed_l2=0.0, layer_l2=0.0)
model.run_eval(test_file, num_ngs=9)   # -> {"auc": ..., "group_auc": ..., ...}
```

### New files

- `recommenders/models/deeprec/models/sequential/pytorch/rnn_cell_pytorch.py` — `Time4LSTMCell` + a `dynamic_rnn`-equivalent scan driver.
- `recommenders/models/deeprec/models/sequential/pytorch/sequential_base_pytorch.py` — shared embeddings, unmasked ASVD `Attention`, `FcnNet` MLP head, softmax pairwise loss, unique-embedding regularization, and the `fit`/`run_eval`/`predict` lifecycle.
- `recommenders/models/deeprec/models/sequential/pytorch/sli_rec_pytorch.py` — `SLI_RECModel` (`forward` == the TF `_build_seq_graph`) + masked `AttentionFcn`.
- `recommenders/models/deeprec/io/sequential_dataset_pytorch.py` — faithful port of `SequentialIterator` (parsing, time features, padding, in-batch negatives).
- `tests/unit/recommenders/models/test_sli_rec_pytorch.py` — 12 plain-function unit tests.
- `examples/00_quick_start/slirec_amazon.ipynb` — a **dedicated** PyTorch SLi-Rec quickstart that defines parameters as plain Python and passes them explicitly (no `prepare_hparams`), so the functional test still exercises SLi-Rec end-to-end.

### Removed (TensorFlow SLi-Rec)

- `recommenders/models/deeprec/models/sequential/sli_rec.py` and its TF-only time-aware cell file `rnn_cell_implement.py` (used exclusively by TF SLi-Rec).
- The TF SLi-Rec unit (`test_slirec_component_definition`) and smoke (`test_model_slirec`) tests and their `test_groups.yml` entries.
- `docs/models.rst`: the RNN Cells section; the SLIRec section now points at the PyTorch modules.

### Changed

- `tests/conftest.py` — the `slirec_quickstart` notebook fixture points at the new `slirec_amazon.ipynb`; README links updated.
- `examples/00_quick_start/sequential_recsys_amazondataset.ipynb` — kept as the shared multi-model quickstart, but its default active model is switched from the removed TF SLi-Rec to **A2SVD** (same IJCAI'19 paper), with a pointer to the new SLi-Rec notebook. The other TF sequential models are unaffected.

## Validation

Component-level parity (weights copied TF→PyTorch, on the real slirec data):

- **Time4LSTMCell + scan**: single-step `m`/`c` diff ~1e-7; full padded-sequence `rnn_outputs` diff ~6e-8; padded steps exactly zero.
- **SequentialDataset**: parser, eval-batch arrays, and train-batch arrays (incl. in-batch negative sampling under a shared RNG seed) all bit-identical to the TF iterator.

End-to-end training (10 epochs, batch 400, seed 42, `embed_l2=layer_l2=0` — the functional-test config):

| | untrained | epoch 1 (valid) | epoch 10 (valid) | **test** |
|---|---|---|---|---|
| **PyTorch** AUC | 0.4802 | 0.5071 | 0.7614 | **0.7355** |
| TF AUC | 0.4857 | 0.4975 | 0.7369 | 0.7174 |

The functional test asserts `auc == pytest.approx(0.7183, rel=0.1, abs=0.05)` → accepts `[0.6465, 0.7901]`. The PyTorch test AUC **0.7355** passes comfortably and exceeds both the 0.7183 target and the TF reference (group_auc 0.7201 vs TF 0.7073). Unit tests: 12/12 pass on CPU. The untrained baseline AUC (0.4802) is bit-identical to the earlier `hparams`-based version, confirming the explicit-params refactor preserves initialization and behavior.

## Related Issues

- https://github.com/recommenders-team/recommenders/issues/2302

## Checklist

- [x] I have not rewritten tests or code to pass CI without addressing the underlying issue.
- [x] I ran `black` to format the code.
- [x] I have added tests (unit tests + the existing functional notebook test now exercises the PyTorch model).
- [x] The five other TF sequential models are untouched (byte-identical to staging); their tests still collect.

### CODE REVIEW

File: examples/00_quick_start/sequential_recsys_amazondataset.ipynb [hunk 1]
```diff
@@ -112,7 +111,7 @@
                 "data_path = os.path.join(\"..\", \"..\", \"tests\", \"resources\", \"deeprec\", \"slirec\")\n",
                 "\n",
                 "##  ATTENTION: change to the corresponding config file, e.g., caser.yaml for CaserModel, sum.yaml for SUMModel\n",
-                "yaml_file = '../../recommenders/models/deeprec/config/sli_rec.yaml'  "
+                "yaml_file = '../../recommenders/models/deeprec/config/asvd.yaml'  "
             ]
```
Comment (anargyri) [inline examples/00_quick_start/sequential_recsys_amazondataset.ipynb]: I think there is still SLI-REC in the table with the results. Also need to run with ASVD and add the outputs.

File: examples/00_quick_start/slirec_amazon.ipynb
Comment (anargyri) [inline examples/00_quick_start/slirec_amazon.ipynb:1]: You need to remove the results obtained with the other models.

File: examples/00_quick_start/sequential_recsys_amazondataset.ipynb (hunk 1 above)
Comment (miguelgfierro) [reply examples/00_quick_start/sequential_recsys_amazondataset.ipynb]: fixed

File: examples/00_quick_start/slirec_amazon.ipynb
Comment (miguelgfierro) [reply examples/00_quick_start/slirec_amazon.ipynb:1]: fixed

**Comment by miguelgfierro** [comment]
@anargyri I updated the code, please take a look

File: tests/unit/recommenders/models/test_sli_rec.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        if b
+    )
+    batch = model._to_tensors(np_batch)
+    with torch.no_grad():
+        logit = model.forward(batch)
+    n = np_batch["labels"].shape[0]
+    assert logit.shape == (n, 1)
+
+
+def test_slirec_fit_and_eval_smoke(synthetic_slirec):
```
Comment (anargyri) [inline tests/unit/recommenders/models/test_sli_rec.py:314]: Are these methods used anywhere?

Comment (miguelgfierro) [reply tests/unit/recommenders/models/test_sli_rec.py:314]: Yes — all of them are components of the new PyTorch implementation: `Time4LSTMCell` / `time4lstm_scan` run the time-aware RNN inside `forward` (`pytorch/sli_rec.py`), `Attention` is the long-term ASVD block, `AttentionFcn` the short-term masked attention, and `FcnNet` is both the logit head (`pytorch/sequential_base.py`) and the alpha gate. The public methods are exercised too: `fit` calls `run_eval` internally, and `run_eval` / `predict` / `load_model` are used in the quickstart notebook and smoke tests. The unit tests target these blocks individually to verify each one against the TF reference equations.

**Comment by miguelgfierro** [comment]
FYI @Leavingseason we are migrating some of the sequential algos to PyTorch

File: tests/unit/recommenders/models/test_sli_rec.py (hunk 2 above)
Comment (miguelgfierro) [reply tests/unit/recommenders/models/test_sli_rec.py:314]: cc @anargyri

Comment (anargyri) [reply tests/unit/recommenders/models/test_sli_rec.py:314]: Sorry, I meant the test functions. It's fine, my mistake, I misread something.

**Review by anargyri — APPROVED** [review]
